### PR TITLE
nextcloud21: init at 21.0.0, set as default version

### PR DESCRIPTION
**Pull request by Ma27**
###### Motivation for this change

<del>:warning: This is marked as draft on purpose since [a 21.0.0 tag exists](https://github.com/nextcloud/server/releases/tag/v21.0.0), but no changelog is available at nextcloud.com. Since Nextcloud is planning an "announcement" on 22 Feb, I guess that it's not official yet.</del>

<del>However I decided to give it a try (successfully btw), so I decided to open this PR already :) </del>

------

ChangeLog: https://nextcloud.com/changelog/#latest21

* Packaged 21.0.0, test-deployed it to my personal instance and tested
  the most basic functionality (`davfs2`-mount, {card,cal}dav sync, file
  management).

* Bumped the default version for unstable/21.05 to `nextcloud21`. Since
  `nextcloud20` was added after the release of 20.09 (and thus the
  default on 20.09 is still `nextcloud19`), it's now needed to upgrade
  across two majors.

  This is not a problem though since it's possible to upgrade to v20 on
  20.09 already and if not, the module will guard the administrator
  through the upgrade with eval warnings as it's the case since 20.03.

* Dropped `nextcloud17` attribute and marked `nextcloud18` as EOL.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 113958` at 7f0dfc15 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li><a href="https://gist.github.com/301c6e75228c18c438ef0df970e817dd">nextcloud21</a></li></li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/servers/nextcloud/default.nix:18:5:
```
   |
18 |     installPhase = ''
   |     ^
```
</li>
  </ul>
</details>

**Comment by Ma27** [comment]
cc @bachp @mweinelt @aanderse @xfix @lukegb does anybody of you have some time to test/review this? :)

**Comment by mweinelt** [comment]
Generally works, but:

There are some warnings regarding your setup.
 - A background job is pending that checks for user imported SSL certificates. Please check back later.
 - Your web server is not properly set up to resolve "/.well-known/webfinger". Further information can be found in the documentation.
 - Your web server is not properly set up to resolve "/.well-known/nodeinfo". Further information can be found in the documentation.
 - Your installation has no default phone region set. This is required to validate phone numbers in the profile settings without a country code. To allow numbers without a country code, please add "default_phone_region" with the respective ISO 3166-1 code ↗ of the region to your config file.
 - The database is missing some indexes. Due to the fact that adding indexes on big tables could take some time they were not added automatically. By running "occ db:add-missing-indices" those missing indexes could be added manually while the instance keeps running. Once the indexes are added queries to those tables are usually much faster.
        Missing index "fs_size" in table "oc_filecache".

```
# nextcloud-occ db:add-missing-indices
Check indices of the share table.
Check indices of the filecache table.
Check indices of the twofactor_providers table.
Check indices of the login_flow_v2 table.
Check indices of the whats_new table.
Check indices of the cards table.
Check indices of the cards_properties table.
Check indices of the calendarobjects_props table.
Check indices of the schedulingobjects table.
Check indices of the oc_properties table.
Done.
```

The missing index isn't created.

**Comment by Ma27** [comment]
Thanks for the detailed feedback! Will look into it :)

**Comment by mweinelt** [comment]
I crossed out two warnings that vanished after a while.

**Comment by Ma27** [comment]
Hmm, will confirm that, but it seems as this is done incrementally by nextcloud-cron.

**Comment by xfix** [comment]
nginx configuration should be updated to match https://docs.nextcloud.com/server/21/admin_manual/installation/nginx.html. In particular, I don't see the following rules in Nextcloud module.

```nginx
    location = / {
        if ( $http_user_agent ~ ^DavClnt ) {
            return 302 /remote.php/webdav/$is_args$args;
        }
    }
```

Microsoft clients aren't handled by Nextcloud module.

```nginx
    location ^~ /.well-known {
        # The following 6 rules are borrowed from `.htaccess`

        location = /.well-known/carddav     { return 301 /remote.php/dav/; }
        location = /.well-known/caldav      { return 301 /remote.php/dav/; }
        # Anything else is dynamically handled by Nextcloud
        location ^~ /.well-known            { return 301 /index.php$uri; }

        try_files $uri $uri/ =404;
    }
```

There is no dynamic handling rule.

File: pkgs/servers/nextcloud/default.nix
Comment (ajs124) [inline pkgs/servers/nextcloud/default.nix:46]: We might want to remove this earlier, as in before the 21.05 branch-off, because it will go EOL soon after.

**Review by ajs124 — COMMENTED** [review]
I haven't tested this, but it LGTM. I probably won't be deploying 21 to any of my instances in the next month or two for reasons, though. So maybe people that actually run and plan on running this should review this.

We might want to package https://github.com/nextcloud/notify_push at some point and add options to the module for it. I might do that when upgrading our instances, although that might only be in a few months.

**Review by mweinelt — APPROVED** [review]
LGTM.

**Comment by Vskilet** [comment]
I started to test this today and I've an error in `nextcloud-setup-start` service :
```
Fatal error: Allowed memory size of 536870912 bytes exhausted (tried to allocate 20480 bytes) in /nix/store/j08fjf3yy6hdfvavrhlvd4bixv7hw8zx-nextcloud-21.0.0/lib/private/AppFramework/Utility/SimpleContainer.php on line 133
```

I haven't not much investigating but I found [this](https://github.com/nextcloud/server/issues/25742) : I'll try to set `--define apc.enable_cli=1`.

**Comment by Ma27** [comment]
This seems to be an open issue. However, my instance is running perfectly fine for a while and older versions are still available. Is there anything else to add, otherwise I'd merge soon :)

**Comment by eyJhb** [comment]
Tested this today, seems to work fine for me (keep in mind I set Nextcloud up yesterday).
Had a initial issue where I tried to pick this commit and use on latest Nixos unstable, that however does not work because of some PHP stuff.

**Comment by Ma27** [comment]
>  Had a initial issue where I tried to pick this commit and use on latest Nixos unstable, that however does not work because of some PHP stuff.

@eyJhb what happened there? May be relevant to take care of before merging.

Regarding your message from IRC: 

> Currently I have to force a refresh, and I am told that should not be required. 

What do I have to do to reproduce?

**Comment by eyJhb** [comment]
@ma27 I think I might have done it wrong. I just used the latest `nixos-unstable-small`, and then set the package to the one in this PR (version 21).

Then I got this out 

```
Mar 03 10:06:18 toffee php-fpm[16193]: [03-Mar-2021 10:06:18] NOTICE: PHP message: PHP Warning:  PHP Startup: Unable to load dynamic library '/nix/store/ixcm7yfd5f6y15k53lmaxra4rb1j76yi-php-7.4.15-imagick-3.4.4/lib/php/extensions/php.so' (tried: /nix/store/ixcm7yfd5f6y15k53lmaxra4rb1j76yi-php-7.4.15-imagick-3.4.4/lib/php/extensions/php.so (/nix/store/ixcm7yfd5f6y15k53lmaxra4rb1j76yi-php-7.4.15-imagick-3.4.4/lib/php/extensions/php.so: cannot open shared object file: No such file or directory), /nix/store/18rr3d4vm65yaxcfisxx8kxjviqyyq8h-php-7.4.15/lib/php/extensions//nix/store/ixcm7yfd5f6y15k53lmaxra4rb1j76yi-php-7.4.15-imagick-3.4.4/lib/php/extensions/php.so.so (/nix/store/18rr3d4vm65yaxcfisxx8kxjviqyyq8h-php-7.4.15/lib/php/extensions//nix/store/ixcm7yfd5f6y15k53lmaxra4rb1j76yi-php-7.4.15-imagick-3.4.4/lib/php/extensions/php.so.so: cannot open shared object file: No such file or directory)) in Unknown on line 0
```

But that might just be how that works.

**Comment by Ma27** [comment]
@eyJhb which rev were you on and what did you do to fix it? Asking because there might be a change on master that isn't tested against this branch and when merging this, Nextcloud would start to fail.

Also, do you still have the problem you described in IRC today? If so, how can I reproduce it?

**Comment by eyJhb** [comment]
> @eyJhb which rev were you on and what did you do to fix it? Asking because there might be a change on master that isn't tested against this branch and when merging this, Nextcloud would start to fail.

Sorry! So the rev I am running it on is `d12574353329291e661805312544b1c68c13911e` and then I pick out this build by doing so

```
{ config, pkgs, ... }:

let
  ma27pkgs = import /home/eyjhb/projects/nixos/nixpkgs {};
in {
  services.nextcloud = {
    package = ma27pkgs.nextcloud21;
   ....
  };
```

I am not sure  if this is a incorrect way of doing it. I just think it "should work" (TM).

But instead I get this page 

```
Internal Server Error

The server encountered an internal error and was unable to complete your request.
Please contact the server administrator if this error reappears multiple times, please include the technical details below in your report.
More details can be found in the server log.
```

I don't really get much in my logs, and I am not entirely sure where they should be located either way (I only have the above warning, but that seems to always be there).

> Also, do you still have the problem you described in IRC today? If so, how can I reproduce it?

Sorry, I completely oversaw this. The issue was actually Deck not supporting this. Opening notes instead, I could see it update in real time.
So I was mislead to believe that Deck should do so as well, but confirmed that it is not the case.

**Comment by beardhatcode** [comment]
I was also going to make an NC21 PR, but I noticed I needed to set `"apc.enable_cli" = "1";` in the phpSettings. Otherwise nextcloud goes out of memory if APC is enabled.

**Comment by mweinelt** [comment]
I've disabled apcu and enabled redis instead. Can't exactly remember why.

**Comment by Ma27** [comment]
So, PHP extension on master are currently broken (see #nixos for details) and hence that's unrelated to this change.

**Comment by Ma27** [comment]
@beardhatcode the module now sets apc.enable_cli if apcu caching is enabled.

Given that the extension issue is now fixed on master, I'll wait for ofborg and merge then.

**Comment by FRidh** [comment]
regression reported in https://github.com/NixOS/nixpkgs/issues/116159

**Pull request merged**
